src/metabase/driver/firebolt.clj

```diff
@@ -238,134 +230,35 @@
-                                            (reduce [_ rf init]
-                                              (reduce
-                                               rf
-                                               init
-                                               (when @has-fields-without-type-info?
-                                                 (fallback-fields-metadata-from-select-query driver conn schema table-name)))))]
-        ;; VERY IMPORTANT! DO NOT REWRITE THIS TO BE LAZY! IT ONLY WORKS BECAUSE AS NORMAL-FIELDS GETS REDUCED,
-        ;; HAS-FIELDS-WITHOUT-TYPE-INFO? WILL GET SET TO TRUE IF APPLICABLE AND THEN FALLBACK-FIELDS WILL RUN WHEN
-        ;; IT'S TIME TO START EVALUATING THAT.
-        (reduce
-         ((comp cat (m/distinct-by :name)) rf)
-         init
-         [jdbc-metadata fallback-metadata])))))
-
-(defn describe-table-fields
-  "Returns a set of column metadata for `table` using JDBC Connection `conn`."
-  [driver conn table & [db-name-or-nil]]
-  (into
-   #{}
-   (map-indexed (fn [i {:keys [database-type], column-name :name, :as col}]
-                  (merge
-                   (u/select-non-nil-keys col [:name :database-type :field-comment])
-                   {:base-type         (database-type->base-type-or-warn driver database-type)
-                    :database-position i}
-                   (when-let [semantic-type (calculated-semantic-type driver column-name database-type)]
-                     {:semantic-type semantic-type}))))
-   (fields-metadata driver conn table db-name-or-nil)))
-
-(defn add-table-pks
-  "Using `metadata` find any primary keys for `table` and assoc `:pk?` to true for those columns."
-  [^DatabaseMetaData metadata table]
-  (let [pks (into #{} (common/reducible-results #(.getPrimaryKeys metadata nil nil (:name table))
-                                                (fn [^ResultSet rs]
-                                                  #(.getString rs "COLUMN_NAME"))))]
-    (update table :fields (fn [fields]
-                            (set (for [field fields]
-                                   (if-not (contains? pks (:name field))
-                                     field
-                                     (assoc field :pk? true))))))))
-
-(defmethod driver/describe-table :firebolt
-  [driver database table]
-  (let [spec (sql-jdbc.conn/db->pooled-connection-spec database)]
-    (with-open [conn (jdbc/get-connection spec)]
-      (->> (assoc (select-keys table [:name :schema])
-                  :fields (describe-table-fields driver conn table nil))
-           ;; find PKs and mark them
-           (add-table-pks (.getMetaData conn))))))
+(defmethod sql-jdbc.describe-table/get-table-pks :firebolt
+  [_ ^Connection conn db-name-or-nil table]
+  (let [table-name (get table :name)
+        schema (get table :schema)
+        base-sql "SELECT primary_index FROM information_schema.tables WHERE table_name = ?"
```

Comment:
Should we be reporting primary index as a primary key? As far as i understand there's an important distinction - PIs are not unique.